### DISCUSSION TITLE
V3.2 connecting and disconnecting devices

### DISCUSSION THREAD

**Discussion by vascolp** [pybricks/support#888] [Q&A]
Hi!

In order to make my [Remote Bla Bla](https://github.com/vascolp/RemoteBlaBla) software work, I need to be able to connect and disconnect motors to the Hub.
Everything was fine in v3.1, but in v3.2 it does not work.
The simple program below fails in the second call to `m=Motor(Port.A)` with an

     `OSError: [Errno 16] EBUSY: Device or resource busy`

Is there a way to fix this?

```python
from pybricks.pupdevices import DCMotor, Motor, Remote
from pybricks.parameters import Port, Direction, Stop, Button, Color
from pybricks.tools import wait

print('Connect motor to port A')
wait(2000)
m=Motor(Port.A)
m.dc(50)
wait(2000)
m.dc(0)
print('disconnect motor')
wait(2000)
print('Connect motor to port A')
wait(2000)
m=Motor(Port.A)
m.dc(100)
wait(2000)
m.dc(0)
print('bye')
wait(2000)
```

Thanks in advance!
Vasco

**Comment by dlech** [comment]
Can you just remove the second `m=Motor(Port.A)`? The `m.dc()` command after that should fail if the motor is not connected.

**Reply by vascolp** [reply]
No... the real code is not as simple as that...  It is Motor for motors with sensors but it is DCMotor for simple motors. There is a call to                 PUPDevice() in between to know the device type connected...
The user can connect any device type, and I have to cope with it.

**Reply by dlech** [reply]
I think we would need to add a `Motor.close()` method that frees the resources to support this use case.

**Reply by vascolp** [reply]
By the way, this is a configuration step, so the user must be able to connect a device, test if it does what he wants, and if it doesnt, he can disconnect and connect another device. So it must be possible to disconnect and connect devices with no problems.

**Reply by vascolp** [reply]
Yes Motor.close() or in fact any_device.close() should do the job! 
But you don´t have that, do you?  :-(

**Reply by dlech** [reply]
No, not currently.

**Reply by vascolp** [reply]
Is there a workaround for that? Or is the data at C level and I cannot do anything, even if "illegal"?
Because in previous versions it was possible to call Motor several times....

**Reply by dlech** [reply]
It's all at the C level, so the closest thing to a workaround is stopping the program and starting it again.

**Reply by vascolp** [reply]
Can I do that from code? I supose not, and this works connected to the remote, so I would have to ask the user to reconnect the remote... really not good....

**Comment by vascolp** [comment]
I have been trying to cope with this situation but it is impossible for my program. The whole ideia of Remote Bla Bla to be able to configure the devices I want to use by trial and error. That means configure, test, reconfigure. For that I need to connect and reconnect devices.

Can you please consider making a Motor/DCMotor.close() method as soon as possible? Beta is ok for me... Thank you!
